modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/RegistryLookupValueResolverWrapper.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.runtime.resolver;
+
+import static java.lang.String.format;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.mule.runtime.api.i18n.I18nMessageFactory.createStaticMessage;
+import static org.mule.runtime.api.util.Preconditions.checkArgument;
+import static org.mule.runtime.core.api.lifecycle.LifecycleUtils.initialiseIfNeeded;
+
+import org.mule.runtime.api.artifact.Registry;
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.lifecycle.Initialisable;
+import org.mule.runtime.api.lifecycle.InitialisationException;
+import org.mule.runtime.core.api.MuleContext;
+import org.mule.runtime.core.api.config.ConfigurationException;
+import org.mule.runtime.core.api.context.MuleContextAware;
+import org.mule.runtime.core.api.util.func.Once;
+
+import javax.inject.Inject;
+
+/**
+ * Implementation of {@link ValueResolver} which accesses the mule registry and returns the value associated with the result of
+ * another {@link ValueResolver}.
+ * <p/>
+ * Although the registry is mutable, the value associated to a given key is not meant to change. As a consequence,
+ * {@link #isDynamic()} will delegate to the wrapped {@link ValueResolver}.
+ *
+ * @since 4.5.0
+ */
+public class RegistryLookupValueResolverWrapper<T> implements ValueResolver<T>, Initialisable, MuleContextAware {
+
+  private final ValueResolver<String> delegate;
+  private MuleContext muleContext;
+  private Registry registry;
+
+  /**
+   * Construct a new instance and set the {@link ValueResolver} of the key that will be used to access the registry
+   *
+   * @param delegate a resolver for a not blank {@link String}
+   */
+  public RegistryLookupValueResolverWrapper(ValueResolver<String> delegate) {
+    this.delegate = delegate;
+  }
+
+  /**
+   * Returns the registry value associated with the result of the {@link #delegate} {@link ValueResolver}.
+   *
+   * @param context a {@link ValueResolvingContext}
+   * @return the registry value associated with the key resolved from {@link #delegate}
+   * @throws MuleException          if an error occurred fetching the value
+   * @throws ConfigurationException if no object is registered under the key resolved from {@link #delegate}
+   */
+  @Override
+  public T resolve(ValueResolvingContext context) throws MuleException {
+    String key = delegate.resolve(context);
+    checkArgument(!isBlank(key), "A null or empty key was provided. Registry lookup cannot be performed with a blank key");
+
+    return registry.<T>lookupByName(key)
```

Comment:
this should be cached. For any given key, the registry should be accessed only once

Reply:
In the latest version, I added caching in the case where the reference is static.

But for the case of a dynamic reference (expression in config-ref) I think the caching shouldn't be done at the resolver level and would be better placed inside the registry itself. My reasoning is the following, please correct me if I'm wrong:

1. If we put the cache at the ValueResolver **class**, we might have problems with providers of the same name across different apps. Would classloader isolation protect us from this? And also, what about lifecycle on redeployments, I'm worried the cache would retain disposed instances that are no longer valid.
2. If we put the cache at the ValueResolver **instance**, we would probably be wasting extra memory for cases in which the caches for different resolvers have the same contents. And we wouldn't benefit from previous hits on the same key by different resolver instances.
3. If the registry access is expensive, we should place the cache in a common place so that everyone accessing the registry in runtime goes through it, right?
4. Are we sure that accessing the registry is more expensive than accessing a cache map (concurrent hash map most likely)?

Reply:
Caching at the registry level is not that straight forward:

A registry level cache should:

1) Be agnostic of the type of object being handled (no special treatment for configuration providers)
2) consider prototype objects, which are specific to spring registry but don't exist in the SimpleRegistry (used by MuleFramework)
3) Consider that objects can be registered/unregistered at any time

Although there are some merits to your point, caching at the registry level is very complicated yet never been a problem since lookups usually happen at startup time. All the problems you mention disappear if you use the ExtensionManager instead.

As for lifecycle, that's not an issue as configs lifecycle is tied to that of the mule context.